### PR TITLE
Don't logout from iscsi target until all block/filesystem mappers/mounters are done

### LINKED ISSUE #74313: Iscsi raw block support logs out from target too early
On the detach code path for iSCSI raw block volumes, there is no ref count check to avoid logging out too early.

**What happened**:
When kubelet detaches an iSCSI raw block volume, it always logs out of the target, even if there are other pods using other volumes on the same target.

**What you expected to happen**:
Kubelet should not log out of the target until the last volume has been disconnected.

**How to reproduce it (as minimally and precisely as possible)**:
Create 2 raw block PVCs.
Create 2 iscsi PVs and bind them to those PVCs (using a dynamic provisioner, or manually)
Create 2 pods, 1 connected to each PVC, ensuring they land on the same node (either use a single node cluster, or force the pods to a particular node in a multi node cluster)
Verify that the pod are running, an iSCSI session exists, and 2 disks are found on the node where the pods run.
Delete 1 of the pods.
Notice that both block devices are gone, and the iSCSI session is gone, and the remaining pod gets I/O errors on its block device.

**Anything else we need to know?**:
The problem also affects ordinary non-raw-block (filesystem) volumes on the same iSCSI target. When a pod containing a raw block volume is deleted, the filesystem volumes are detached too.

**Environment**:
- Kubernetes version: v1.15.0-alpha.0.81+82b7a8a5e05a57
- Cloud provider or hardware configuration: qemu VM
- OS: Arch Linux 2019.02.01
- Kernel: Linux dev1 4.20.7-arch1-1-ARCH #1 SMP PREEMPT Wed Feb 6 18:42:40 UTC 2019 x86_64 GNU/Linux
- Install tools: kubernetes-the-hard-way
- Others: n/a

### PR DESCRIPTION
**Pull request by wongma7**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
- Block volume detach always logs out. This disrupts other block/file users
- Filesystem volume detach logs out even if there are other block users.
  - Fix: count both block users and file users in `getDevicePrefixRefCount`

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubernetes/issues/74313

**Special notes for your reviewer**: iscsi expertise needed
TODO: tests

**Does this PR introduce a user-facing change?**:
```release-note
TODO
```

### CODE REVIEW

File: pkg/volume/iscsi/iscsi_util.go [hunk 1]
```diff
@@ -695,31 +733,33 @@ func (util *ISCSIUtil) DetachBlockISCSIDisk(c iscsiDiskUnmapper, mapPath string)
...
-		}
-		lun = arr[1]
 		// Extract the iface from the mountPath and use it to log out. If the iface
 		// is not found, maintain the previous behavior to facilitate kubelet upgrade.
 		// Logout may fail as no session may exist for the portal/IQN on the specified interface.
 		iface, found = extractIface(mapPath)
 	}
+
+	// Delete all the scsi devices and any multipath devices
+	if err = deleteDevices(c.iscsiDisk, c.exec, c.deviceUtil); err != nil {
```
Comment (wongma7) [inline pkg/volume/iscsi/iscsi_util.go]: the block detach path did not call this `deleteDevices` before. Why? I am pretty sure it should.

File: pkg/volume/iscsi/iscsi_util.go [hunk 2]
```diff
@@ -695,31 +733,33 @@ func (util *ISCSIUtil) DetachBlockISCSIDisk(c iscsiDiskUnmapper, mapPath string)
...
+	}
+
 	portals := removeDuplicate(bkpPortal)
 	if len(portals) == 0 {
 		return fmt.Errorf("iscsi detach disk: failed to detach iscsi disk. Couldn't get connected portals from configurations")
 	}
 
-	devicePath := getDevByPath(portals[0], iqn, lun)
-	klog.V(5).Infof("iscsi: devicePath: %s", devicePath)
-	if _, err = os.Stat(devicePath); err != nil {
```
Comment (wongma7) [inline pkg/volume/iscsi/iscsi_util.go]: the block detach path tries to "validate devicePath" but the filesystem does not. Why might it be needed? I don't think it is, so I am removing it.

File: pkg/volume/iscsi/iscsi_util.go (hunk 1 above)
Comment (wongma7) [reply pkg/volume/iscsi/iscsi_util.go]: After this change and the below removal of os.Stat(devicePath), `DetachBlockISCSIDisk` and `DetachDisk` will look almost identical except `DetachDisk` tries to unmount the mntPath

File: pkg/volume/iscsi/iscsi_util.go [hunk 3]
```diff
@@ -665,26 +699,30 @@ func (util *ISCSIUtil) DetachDisk(c iscsiDiskUnmounter, mntPath string) error {
...
 		return err
 	}
+
 	var bkpPortal []string
-	var volName, iqn, lun, iface, initiatorName string
+	var volName, iqn, iface, initiatorName string
 	found := true
+
 	// load iscsi disk config from json file
+	// TODO: if the disk was first attached as filesystem then detached as block, the iscsi.json will not be under volumeDevices, vice-versa
```
Comment (wongma7) [inline pkg/volume/iscsi/iscsi_util.go]: Not sure if this is a problem, need to double-check

File: pkg/volume/iscsi/iscsi_util.go [hunk 4]
```diff
@@ -695,31 +733,33 @@ func (util *ISCSIUtil) DetachBlockISCSIDisk(c iscsiDiskUnmapper, mapPath string)
...
 		return fmt.Errorf("iscsi detach disk: failed to detach iscsi disk. Couldn't get connected portals from configurations")
 	}
 
-	devicePath := getDevByPath(portals[0], iqn, lun)
-	klog.V(5).Infof("iscsi: devicePath: %s", devicePath)
-	if _, err = os.Stat(devicePath); err != nil {
-		return fmt.Errorf("failed to validate devicePath: %s", devicePath)
-	}
-	// check if the dev is using mpio and if so mount it via the dm-XX device
-	if mappedDevicePath := c.deviceUtil.FindMultipathDeviceForDevice(devicePath); mappedDevicePath != "" {
```
Comment (redbaron) [inline pkg/volume/iscsi/iscsi_util.go]: this removes use of a multipath, IMHO we should keep it

**Comment by wongma7** [comment]
@kubernetes/sig-storage-pr-reviews @kubernetes/sig-storage-bugs this is ready for review.

iscsi tests are behind Feature:Volumes so will not run in CI. Tested locally: with patch, tests pass.
I built the image locally by doing:
`cd test/images/volumes/iscsi; docker build . -t gcr.io/kubernetes-e2e-test-images/volume/iscsi:1.1` and ran `go run hack/e2e.go -v --test --provider=local --kubeconfig=/var/run/kubernetes/admin.kubeconfig --test_args='--ginkgo.focus=\[Driver\:\siscsi\].*should\saccess\san\siSCSI'`

Ran 4 of 4331 Specs in 766.583 seconds
SUCCESS! -- 4 Passed | 0 Failed | 0 Pending | 4327 Skipped PASS

Without patch (master), 3 tests fail (fs+block, block+fs, block+block) & fs+fs passes. "fs+block" means pod0 is fs and pod1 is block, we delete pod1 while pod0 is still running then check if pod0 can read/write the volume.

- fs+block: expect this to fail because on master, deleting the block volume pod ALWAYS logs out
```
[sig-storage] In-tree Volumes [Driver: iscsi][Feature:Volumes] [Testpattern: Pre-provisioned PV (filesystem volmode)] multiVolume
  should access an iSCSI volume after a different pod on the same node using a volume with a different volume mode, same target, & different lun is deleted
```
- block+block: same reasoning as above
```
[sig-storage] In-tree Volumes [Driver: iscsi][Feature:Volumes] [Testpattern: Pre-provisioned PV (block volmode)] multiVolume
  should access an iSCSI volume after a different pod on the same node using a volume with the same volume mode, same target, & different lun is deleted
```
- block+fs: expect this to fail because on master, deleting the fs volume pod does not check for existence of block volume pods
```
[sig-storage] In-tree Volumes [Driver: iscsi][Feature:Volumes] [Testpattern: Pre-provisioned PV (block volmode)] multiVolume
  should access an iSCSI volume after a different pod on the same node using a volume with a different volume mode, same target, & different lun is deleted
```
- fs+fs: expect this to succeed because on master, logout checks for mounts. This is legacy behaviour.
```
[sig-storage] In-tree Volumes [Driver: iscsi][Feature:Volumes] [Testpattern: Pre-provisioned PV (filesystem volmode)] multiVolume
  should access an iSCSI volume after a different pod on the same node using a volume with the same volume mode, same target, & different lun is deleted
```

**Comment by wongma7** [comment]
commit 3 is cherry-picked from https://github.com/kubernetes/kubernetes/pull/74693.

**Comment by jsafrane** [comment]
cc: @bswartz @j-griffith for iSCSI expertise

**Comment by bswartz** [comment]
This looks promising, but it's a huge amount of code. I will download the patch and test it before reading all the chnges

**Comment by bswartz** [comment]
Okay I set up some tests to exercise the failure modes I know about, and this patch does prevent the premature logouts, but it creates new problems where the logouts no longer happen when they should.

After creating 4 pods with 1 PVC each (2 filesystem, 2 block) and then deleting those 4 pods, I end up with 2 disks still attached (the block ones) and the iSCSI session still logged in. I haven't delved into the code yet to see why this might be happening, but this will need fixing before we can merge the bugfix.

**Comment by wongma7** [comment]
Thank you for trying it, I think you've found that there is room for races when multiple unmounts/unmaps happen simultaneously, which I neglected to test for. There needs to be some synchronization within the plugin surrounding the `check for other unmaps/unmounts` and `do unmap/unmount` operations, otherwise all the volumes will see each other then none of them will log out. There's already a lock but it's evidently misplaced...

**Comment by wongma7** [comment]
/hold
need to fix the sync/race issues and add more tests.

**Comment by bswartz** [comment]
I'm not sure it's a race causing this problem. I only ever deleted 1 pod at a time, and pod deletion was by far the slowest part of the tests.

Also it's the block volumes that aren't getting detached, not the filesystem volumes, so I don't think mount/unmount has anything to do with it.

I suspect that we're either missing logic to detach individual block volumes (by writing "1" to /sys/block/sdX/device/delete) or else that logic isn't triggering, and the failure to detach the individual volumes is causing the failure to logout.

I'll take a look at the code soon, but the sheer size of the PR means that will be a big task.

**Comment by wongma7** [comment]
ok I see, that is extremely suspect & perplexing because master code *doesn't* do that detach (`deleteDevice`) for block volumes but my PR *does* because I guessed it should: https://github.com/kubernetes/kubernetes/pull/74627/files#diff-50e99d9d6e7992429dd502362f1491afR754 .

re: the size of the PR, you can ignore the 3rd commit which makes up most of the line changes. Thanks again.

**Comment by wongma7** [comment]
@bswartz I couldn't reproduce the issue, could you share the 4 PV specs?

**Comment by bswartz** [comment]
> @bswartz I couldn't reproduce the issue, could you share the 4 PV specs?

[arch@dev1 ~]$ kubectl get pv -o yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: PersistentVolume
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"v1","kind":"PersistentVolume","metadata":{"annotations":{},"name":"pv1"},"spec":{"accessModes":["ReadWriteOnce"],"capacity":{"storage":"1Gi"},"claimRef":{"apiVersion":"v1","kind":"PersistentVolumeClaim","name":"pvc1","namespace":"default"},"iscsi":{"fsType":"ext4","iqn":"iqn.2002-08.org.swartzlander:iscsid1","lun":1,"readOnly":false,"targetPortal":"192.168.30.21:3260"},"persistentVolumeReclaimPolicy":"Delete","storageClassName":"manual","volumeMode":"Filesystem"}}
    creationTimestamp: "2019-04-02T19:10:26Z"
    finalizers:
    - kubernetes.io/pv-protection
    name: pv1
    resourceVersion: "337"
    selfLink: /api/v1/persistentvolumes/pv1
    uid: f8e2fe87-557a-11e9-9bed-525400003011
  spec:
    accessModes:
    - ReadWriteOnce
    capacity:
      storage: 1Gi
    claimRef:
      apiVersion: v1
      kind: PersistentVolumeClaim
      name: pvc1
      namespace: default
      resourceVersion: "314"
      uid: f8dca0e4-557a-11e9-9bed-525400003011
    iscsi:
      fsType: ext4
      iqn: iqn.2002-08.org.swartzlander:iscsid1
      iscsiInterface: default
      lun: 1
      targetPortal: 192.168.30.21:3260
    persistentVolumeReclaimPolicy: Delete
    storageClassName: manual
    volumeMode: Filesystem
  status:
    phase: Bound
- apiVersion: v1
  kind: PersistentVolume
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"v1","kind":"PersistentVolume","metadata":{"annotations":{},"name":"pv2"},"spec":{"accessModes":["ReadWriteOnce"],"capacity":{"storage":"1Gi"},"claimRef":{"apiVersion":"v1","kind":"PersistentVolumeClaim","name":"pvc2","namespace":"default"},"iscsi":{"fsType":"ext4","iqn":"iqn.2002-08.org.swartzlander:iscsid1","lun":2,"readOnly":false,"targetPortal":"192.168.30.21:3260"},"persistentVolumeReclaimPolicy":"Delete","storageClassName":"manual","volumeMode":"Filesystem"}}
    creationTimestamp: "2019-04-02T19:10:26Z"
    finalizers:
    - kubernetes.io/pv-protection
    name: pv2
    resourceVersion: "340"
    selfLink: /api/v1/persistentvolumes/pv2
    uid: f91c5705-557a-11e9-9bed-525400003011
  spec:
    accessModes:
    - ReadWriteOnce
    capacity:
      storage: 1Gi
    claimRef:
      apiVersion: v1
      kind: PersistentVolumeClaim
      name: pvc2
      namespace: default
      resourceVersion: "318"
      uid: f91b75fd-557a-11e9-9bed-525400003011
    iscsi:
      fsType: ext4
      iqn: iqn.2002-08.org.swartzlander:iscsid1
      iscsiInterface: default
      lun: 2
      targetPortal: 192.168.30.21:3260
    persistentVolumeReclaimPolicy: Delete
    storageClassName: manual
    volumeMode: Filesystem
  status:
    phase: Bound
- apiVersion: v1
  kind: PersistentVolume
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"v1","kind":"PersistentVolume","metadata":{"annotations":{},"name":"pv3"},"spec":{"accessModes":["ReadWriteOnce"],"capacity":{"storage":"1Gi"},"claimRef":{"apiVersion":"v1","kind":"PersistentVolumeClaim","name":"pvc3","namespace":"default"},"iscsi":{"iqn":"iqn.2002-08.org.swartzlander:iscsid1","lun":3,"readOnly":false,"targetPortal":"192.168.30.21:3260"},"persistentVolumeReclaimPolicy":"Delete","storageClassName":"manual","volumeMode":"Block"}}
    creationTimestamp: "2019-04-02T19:10:26Z"
    finalizers:
    - kubernetes.io/pv-protection
    name: pv3
    resourceVersion: "342"
    selfLink: /api/v1/persistentvolumes/pv3
    uid: f955b4e5-557a-11e9-9bed-525400003011
  spec:
    accessModes:
    - ReadWriteOnce
    capacity:
      storage: 1Gi
    claimRef:
      apiVersion: v1
      kind: PersistentVolumeClaim
      name: pvc3
      namespace: default
      resourceVersion: "321"
      uid: f9552327-557a-11e9-9bed-525400003011
    iscsi:
      iqn: iqn.2002-08.org.swartzlander:iscsid1
      iscsiInterface: default
      lun: 3
      targetPortal: 192.168.30.21:3260
    persistentVolumeReclaimPolicy: Delete
    storageClassName: manual
    volumeMode: Block
  status:
    phase: Bound
- apiVersion: v1
  kind: PersistentVolume
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"v1","kind":"PersistentVolume","metadata":{"annotations":{},"name":"pv4"},"spec":{"accessModes":["ReadWriteOnce"],"capacity":{"storage":"1Gi"},"claimRef":{"apiVersion":"v1","kind":"PersistentVolumeClaim","name":"pvc4","namespace":"default"},"iscsi":{"iqn":"iqn.2002-08.org.swartzlander:iscsid1","lun":4,"readOnly":false,"targetPortal":"192.168.30.21:3260"},"persistentVolumeReclaimPolicy":"Delete","storageClassName":"manual","volumeMode":"Block"}}
    creationTimestamp: "2019-04-02T19:10:27Z"
    finalizers:
    - kubernetes.io/pv-protection
    name: pv4
    resourceVersion: "346"
    selfLink: /api/v1/persistentvolumes/pv4
    uid: f9847552-557a-11e9-9bed-525400003011
  spec:
    accessModes:
    - ReadWriteOnce
    capacity:
      storage: 1Gi
    claimRef:
      apiVersion: v1
      kind: PersistentVolumeClaim
      name: pvc4
      namespace: default
      resourceVersion: "324"
      uid: f983f6d2-557a-11e9-9bed-525400003011
    iscsi:
      iqn: iqn.2002-08.org.swartzlander:iscsid1
      iscsiInterface: default
      lun: 4
      targetPortal: 192.168.30.21:3260
    persistentVolumeReclaimPolicy: Delete
    storageClassName: manual
    volumeMode: Block
  status:
    phase: Bound
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
[arch@dev1 ~]$

**Comment by jsafrane** [comment]
@wongma7, @bertinatto, is this the same as #80091?

**Comment by bertinatto** [comment]
> @wongma7, @bertinatto, is this the same as #80091?

Yes, I think we can close this one if @wongma7 agrees.

**Comment by wongma7** [comment]
yes, TY for taking it

**Pull request closed without merging**